schema/avro/serde.go

```diff
@@ -0,0 +1,98 @@
+// Copyright © 2023 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package avro
+
+import (
+	"fmt"
+
+	"github.com/hamba/avro/v2"
+)
+
+// Serde represents an Avro schema. It exposes methods for marshaling and
+// unmarshalling data.
+type Serde struct {
+	schema        avro.Schema
+	unionResolver unionResolver
+}
+
+// Marshal returns the Avro encoding of v. Note that this function may mutate v.
+// Limitations:
+// - Map keys need to be of type string,
+// - Array values need to be of type uint8 (byte).
+func (s *Serde) Marshal(v any) ([]byte, error) {
+	err := s.unionResolver.BeforeMarshal(v)
+	if err != nil {
+		return nil, err
+	}
+	bytes, err := avro.Marshal(s.schema, v)
+	if err != nil {
+		return nil, fmt.Errorf("could not marshal into avro: %w", err)
+	}
+	return bytes, nil
+}
+
+// Unmarshal parses the Avro encoded data and stores the result in the value
+// pointed to by v. If v is nil or not a pointer, Unmarshal returns an error.
+// Note that arrays and maps are unmarshalled into slices and maps with untyped
+// values (i.e. []any and map[string]any). This is a limitation of the Avro
+// library used for encoding/decoding the payload.
+func (s *Serde) Unmarshal(b []byte, v any) error {
+	err := avro.Unmarshal(s.schema, b, v)
+	if err != nil {
+		return fmt.Errorf("could not unmarshal from avro: %w", err)
+	}
+	err = s.unionResolver.AfterUnmarshal(v)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// String returns the canonical form of the schema.
+func (s *Serde) String() string {
+	return s.schema.String()
+}
+
+// Sort fields in the schema. It can be used in tests to ensure the schemas can
+// be compared.
+func (s *Serde) Sort() {
+	traverseSchema(s.schema, sortFn)
+}
+
+// Parse parses a schema string.
+func Parse(text string) (*Serde, error) {
+	schema, err := avro.Parse(text)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse avro schema: %w", err)
+	}
+	return &Serde{
```

Comment:
Is there a specific reason for not sorting the fields here (as is done in SerdeForType)?

Reply:
Yes, the order plays a role in Avro schemas, see for instance how [records get encoded](https://avro.apache.org/docs/1.8.2/spec.html#binary_encode_complex):

> A record is encoded by encoding the values of its fields in the order that they are declared.

So sorting here would change the schema and the output. Why can we do it in `SerdeForType`? Well, there we don't start with a schema but rather with a value and then create a schema out of it and we want the created schema to be consistent.

This reminds me that we shouldn't expose the `Sort` method as it can lead to bugs. If a schema is created using `SerdeForType`, the created schema will be already sorted, while if it's created using `Parse` it shouldn't be sorted, as it wouldn't necessarily match the original one passed to `Parse`.

Reply:
Sounds good to me. It would be good to have the above as a comment in the code itself, it will be a nice reminder whenever we revisit the code again.

Reply:
Done 👍